server/app/query/servers.tsx

```diff
@@ -122,13 +127,45 @@ export class RemoteServer {
   ): WebSocket {
     const ws = this.logsSocket(id);
     ws.onmessage = (event) => {
-      const newLog: ServerLog = {
-        remoteServer: this,
-        logLine: event.data,
-      };
-
-      // only retain last 1000 logs
-      setLogs((prevLogs) => [...prevLogs.slice(-1000), newLog]);
+      let newLog: ServerLog;
+      try {
+        const logValue = JSON.parse(event.data);
+        newLog = {
+          remoteServer: this,
+          logLine: logValue.record.message,
+          timestamp: logValue.record.time.timestamp,
+        };
+      } catch (e) {
+        newLog = {
+          remoteServer: this,
+          logLine: event.data,
+          timestamp: Date.now(),
+        };
+      }
+
+      // only retain last 10,000 logs
+      const maxNumLogs = 10000;
+      setLogs((prevLogs) => {
```

Comment:
@akoshelev this is where the ordering happens. if the new log is already the last, it is just appended. if not, it uses `lastIndexOf` to find the right spot. this should be O(n) and I would imagine starts from the end of the list, where it will most likely find the right position. 

the other performance knob here is the max lines kept. I upped it to 10k, but we could lower that now that the download option is available.

Reply:
is it server side or client side?

Reply:
client side. client connects directly via websocket to logs from every helper. 

I don't think it would be possible to do server side while they query is running, because a stale log from one helper could always arrive after other logs have reached the client.